### PR TITLE
Added contents for importVolAsClone, cloneOf, virtualCopyOf and volumeGroup

### PR DESCRIPTION
**Pull request by ansarars**

### CODE REVIEW

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 1]
```diff
@@ -154,6 +166,47 @@ During the import volume process, any legacy (non-container volumes) or existing
...
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| accessProtocol     | fc or iscsi  | The access protocol to use when accessing the persistent volume. |
+| virtualCopyOf      | Text         | The name of existing `PersistentVolumeClaim` to be snapped |
+
+!!! important
+    **No other parameters** are required in the `StorageClass` when snapping of a volume outside of those parameters listed in the table above.
+
+### Applying QOS Rules (volumeGroup)
+In the HPE 3PAR or Primera Storage system, the QoS rules are applied to a volume group (volumeset in HPE 3PAR or Primera). To apply QoS rules to a `PersistentVolumeClaim`, one has to create a volume set in the HPE 3PAR or Primera, apply QoS rules to that volumeset and use the **volumeGroup** parameter while creating a `PersistentVolumeClaim`.
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:203]: volumeset is incorrect usage of volume set.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 2]
```diff
@@ -154,6 +166,47 @@ During the import volume process, any legacy (non-container volumes) or existing
...
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| accessProtocol     | fc or iscsi  | The access protocol to use when accessing the persistent volume. |
+| virtualCopyOf      | Text         | The name of existing `PersistentVolumeClaim` to be snapped |
+
+!!! important
+    **No other parameters** are required in the `StorageClass` when snapping of a volume outside of those parameters listed in the table above.
+
+### Applying QOS Rules (volumeGroup)
```
[the remark this answers is not in the document]
Comment (ansarars) [reply docs/container_storage_provider/hpe_3par_primera/index.md:190]: Right now volumeGroup parameter name changes in CSP code is in progress. Will update this section once code is merge to master.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 3]
```diff
@@ -152,7 +156,44 @@ During the import volume process, any legacy (non-container volumes) or existing
...
+During snapshotting process, any existing `PersistentVolumeClaim` defined in the `virtualCopyOf` parameter within a `StorageClass`, will be snapped as `PersistentVolumeClaim` and exposed through the HPE CSI Driver and made available to the Kubernetes cluster.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| accessProtocol     | fc or iscsi  | The access protocol to use when accessing the persistent volume. |
+| virtualCopyOf      | Text         | The name of existing `PersistentVolumeClaim` to be snapped |
+
+!!! important
+    • **No other parameters** are required in the `StorageClass` when snapping of a volume outside of those parameters listed in the table above<br />
+    • Snapshotting using `virtualCopyOf` is independent of snapshot crd availability on kubernetes and it can be performed on any supported kubernetes version<br />
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:187]: Highlight and Capitalize `CRD`, Capitalize first letter of Kubernetes. End bullet list with periods.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 4]
```diff
@@ -152,7 +156,44 @@ During the import volume process, any legacy (non-container volumes) or existing
...
+
+| Parameter        | Option  | Description |
+| ---------------- | ------- | ----------- |
+| cloneOf          | Text    | The name of the `PersistentVolumeClaim` to be cloned. `cloneOf` and `importVolAsClone` are mutually exclusive. |
+| importVolAsClone | Text    | The name of the Primera or HPE 3PAR volume to clone and import. `importVolAsClone` and `cloneOf` are mutually exclusive. |
+| accessProtocol     | fc or iscsi  | The access protocol to use when accessing the persistent volume. |
+
+!!! important
+    • **No other parameters** are required in the `StorageClass` while cloning outside of those parameters listed in the table above<br />
+    • Cloning using above parameters is independent of snapshot crd availability on kubernetes and it can be performed on any supported kubernetes version<br />
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:174]: Highlight and Capitalize `CRD`, Capitalize first letter of Kubernetes. End bullet list with periods.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 5]
```diff
@@ -152,7 +156,44 @@ During the import volume process, any legacy (non-container volumes) or existing
...
+### Snapshotting a volume
+During snapshotting process, any existing `PersistentVolumeClaim` defined in the `virtualCopyOf` parameter within a `StorageClass`, will be snapped as `PersistentVolumeClaim` and exposed through the HPE CSI Driver and made available to the Kubernetes cluster.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| accessProtocol     | fc or iscsi  | The access protocol to use when accessing the persistent volume. |
+| virtualCopyOf      | Text         | The name of existing `PersistentVolumeClaim` to be snapped |
+
+!!! important
+    • **No other parameters** are required in the `StorageClass` when snapping of a volume outside of those parameters listed in the table above<br />
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:186]: when shapshotting a volume

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 6]
```diff
@@ -84,7 +84,11 @@ These parameters are used for volume provisioning and supported platforms.
 |                                     | full    | Indicates Full provisioned volume type. | **X** |   |
 |                                     | dedup   | Indicates Thin Deduplication volume type. | **X** |   |
 |                                     | reduce  | Indicates Thin Deduplication/Compression volume type. |   | **X** |
-| importVol <br /> this is the only option required for importing volume    | Text      | Name of the volume to import via the StorageClass | **X** | **X** |
+| importVol   | Text      | Name of the volume to import via the StorageClass | **X** | **X** |
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:87]: remove "via the StorageClass". This is a table of StorageClass parameters therefore it is already assumed these parameters will be part of a Storageclass

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 7]
```diff
@@ -84,7 +84,11 @@ These parameters are used for volume provisioning and supported platforms.
...
+| importVolAsClone  | Text      | Name of the volume to clone and import via the StorageClass | **X** | **X** |
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:88]: see above

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 8]
```diff
@@ -84,7 +84,11 @@ These parameters are used for volume provisioning and supported platforms.
...
+| cloneOf  | Text      | Name of the `PersistentVolumeClaim` to clone via the StorageClass | **X** | **X** |
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:89]: see above

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 9]
```diff
@@ -84,7 +84,11 @@ These parameters are used for volume provisioning and supported platforms.
...
+| virtualCopyOf  | Text      | Name of the `PersistentVolumeClaim` to snapshot via the StorageClass | **X** | **X** |
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:90]: see above

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (c-snell) [reply docs/container_storage_provider/hpe_3par_primera/index.md:190]: Can you remove the QoS settings section so we can merge this to master. Then we can add QoS once finalized in a new PR.

[the remark this answers is not in the document]
Comment (ansarars) [reply docs/container_storage_provider/hpe_3par_primera/index.md]: volumeGroup parameter is replaced by qosName in StorageClass and code is merged in CSP master, please review this section as well

**Comment by ansarars** [comment]
> Can you remove QoS section so we can merge updates? Once QoS is finalized and through testing submit new PR for QoS parameters.
volumeGroup parameter is replaced with qosName in StorageClass and code is merged in CSP master, please review this section as well

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 10]
```diff
@@ -152,7 +156,44 @@ During the import volume process, any legacy (non-container volumes) or existing
...
+    • **No other parameters** are required in the `StorageClass` when snapshotting a volume outside of those parameters listed in the table above.<br />
+    • Snapshotting using `virtualCopyOf` is independent of snapshot `CRD` availability on Kubernetes and it can be performed on any supported Kubernetes version.<br />
+    • Support for `virtualCopyOf` is available from HPE CSI Driver 1.3.0.
+
+### Applying QOS Rules (qosName)
+In the Primera or HPE 3PAR Storage system, the QoS rules are applied to a volume group (volume set in Primera or HPE 3PAR). To apply QoS rules to a `PersistentVolumeClaim`, one has to create a volume set in the Primera or HPE 3PAR, apply QoS rules to that volume set and use the `qosName` parameter while creating a `PersistentVolumeClaim`.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| qosName      | Text         | Name of the volume set which has QoS rules. This parameter is optional. If specified, the `PersistentVolumeClaim` will be associated with the volume set, for purposes of applying the QoS rules. |
```
Comment (c-snell) [inline docs/container_storage_provider/hpe_3par_primera/index.md:195]: HPE Primera or 3PAR volume set.

**Pull request merged**